### ISSUE TITLE
Load Core Assets from a CDN

### ISSUE THREAD

**Issue by jpdevries** [modxcms/revolution#12647]
This has been discussed in the [Why compress_css and compress_js are silly post](http://forums.modx.com/thread/91496/why-compress-css-and-compress-js-are-silly) forum post and https://github.com/modxcms/revolution/pull/12611 and thus targets 2.5.

Loading common assets such as the core ExtJS files from a CDN would make MODX a more optimized product across the web as a whole. Imagine if logging into your first MODX site of the day meant caching the majority of the assets for the any MODX sites (of the same version) you log into that day. That week. That month, or even ever. Imagine if even the core assets for the MODX sites you haven't even installed yet were in your cache :rocket: 

We would need to have the appropriate files hosted on a CDN with far future reaching expiry dates, GZIP compression, the appropriate headers, and probably to be cached-busted with the MODX version in their file name. I assume there may be cost associated with the CDN hosting? I'm not sure if we can get that sponsored or what.

As I believe @adamwintle mention we would need to do testing and support for countries that block CDNs like China. Maybe there is a `load_cdn_js` System Setting that Defaults to false. It would also be ideal to use the HTML5 Boilerplate trick of attempting to load from a CDN, immediately detecting if the asset was loaded, and falling back to a local copy if needed.

**Comment by pixelchutes** [comment]
What's not to love about this idea? :beers: 

:+1:

**Comment by adamwintle** [comment]
This sounds like a good idea, but I'd definitely make it opt-in and set all defaults to false.

From working in China and other territories that try to block connections, I've noticed that even with a fallback script in place the local copy still never gets loaded. With China for example their firewall will have some CDNs as blocked but then a browser ends up just continually attempt to load the asset, but never hard fail, which is what a fallback script is looking for. 

For example, this code from the [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate/blob/master/dist/index.html#L25): 
```
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="js/vendor/jquery-1.11.3.min.js"><\/script>')</script>
```
Since Google is blocked in China, the Google hosted jQuery libraries are no good; but this file would continually attempt to be loaded, almost as if it's infinitely loading very, very slowly, but never actually end up failing. This then means the `window.jQuery` never gets a `FALSE` response (meaning jQuery wasn't loaded), and the local version is never served. 

From my experience it's always been best to have a self-hosted version on a server in China or use a China-based CDN. All the other workarounds and fallback scripts fail at some point.

**Comment by Jako** [comment]
Automatically included CDN files are never a good idea in a CMS backend. The CDN will track the usage and not everyone wants to use this 'feature'.

**Comment by sottwell** [comment]
How often has some site hung because some CDN or other external source wasn't responding?

**Comment by jpdevries** [comment]
> Automatically included CDN files are never a good idea in a CMS backend. The CDN will track the usage and not everyone wants to use the 'feature' to be tracked.

I agree it should default to false but be available for opt-in users.

**Comment by jpdevries** [comment]
@adamwintle are you aware if cdnjs works in China? I assume not, but curious because it looks like they let people submit libraries so I was thinking of submitting the modx JS and CSS assets for 2.5.

**Comment by adamwintle** [comment]
@jpdevries I did some testing on CDNJS about six months ago, it was intermittently blocked/slow, better than the Google CDN, but definitely not 100% uptime for mainland China users.

**Comment by jpdevries** [comment]
I've written about loading from a CDN in this style guide:
https://github.com/jpdevries/modx-style-guide#lazy-load-scripts

But as @adamwintle points out, it isn't totally reliable. Even if we used local fallbacks as the style guide suggests, CDNs aren't totally reliable. I think we should give users a way to say "don't even try to load stuff from a CDN". Something like a `[[++try_cdn]]` System Setting. People might disable this if they know the CDN could be unreachable or, once we are using service workers, to take advantage of offline support (service workers can only cache stuff on the same domain).


```js
  if(!window.React) promises.push(
    // if needed, try to load React from a CDN (if allowed), fallback to a local copy
    (tryCDN) ? lazyLoadScript(`https://cdnjs.cloudflare.com/ajax/libs/react/${reactVer}/react.min.js`, `react.${reactVer}.min.js`).catch((err => (
                 lazyLoadScript(join(EurekaConfig.assets_url, `./js/vendor/react.${reactVer}.min.js`), `react.${reactVer}.min.js`)
    ))) : lazyLoadScript(join(EurekaConfig.assets_url, `./js/vendor/react.${reactVer}.min.js`), `react.${reactVer}.min.js`)
  );
```

**Comment by OptimusCrime** [comment]
@jpdevries Do we need two issues for this? How is this issue and #13553 different?

**Comment by JoshuaLuckers** [comment]
I don't think loading the core assets via a CDN will ever be implemented. Feel free to reopen this issue or start a new one if I'm wrong.
